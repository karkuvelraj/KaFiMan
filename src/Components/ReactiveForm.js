import React ,{Component}from "react"
import PropTypes from "prop-types"
import {patterns} from "../utils/patterns"
import {  Modal } from "@material-ui/core"
import {default as SeatMap } from "./SeatMap"
import AddBoxIcon from '@material-ui/icons/AddBox';
import { toast } from 'react-toastify';
class ReactiveForm extends Component{

    constructor(props){
        super(props)
        this.state={
             data:this.props.fields.reduce((r,obj)=>{r[obj.id]=obj.value;return r},{}),
             error:{}
                 }
    }
     errDesc={
        mandatory:' is Mandatory',
        email: ' is not a valid email',
        name: ' should not have numbers or special chatacters',
        number: ' is numeric'
    }
   
    handleChange=(event)=>{
        let {type,name,value,required}=event.target
        let err=[]
        const field =this.props.fields.find(x=>x.id==name)
        if(required && !value)err.push(field.name+this.errDesc['mandatory'])
        if(type=='text'&&field.format&&!patterns[field.format].test(value)){
           err.push(field.name+this.errDesc[field.format])
        //    console.log(field.name+this.errDesc[field.format])
        }
        // console.log(err)
        this.setState((prev)=>{
            return {
                data:{
                ...prev.data,
                [name]:value
            },   error:{
                ...prev.error,
                [name]:err
            }
        }
        })
        
    }
    selectModalVal=(value,name)=>{
        this.setState((prev)=>{
            return {
                ...prev,
                data:{
                ...prev.data,
                [name]:value
            }
        }
        })
        this.modalOpenClose('')
    }
    componentDidUpdate(prevProps){

        if(this.props.fields!=prevProps.fields){
            this.setState({
                data:this.props.fields.reduce((r,obj)=>{r[obj.id]=obj.value;return r},{}),
                error:{}
                })
        }
    }
    modalOpenClose=(name)=>{
        this.setState((p)=>{
        return {...p,open:name}
    })}
    checkFields=(event)=>{
        event.preventDefault()
        let flag='';
        this.props.fields.forEach(f => {
         if((this.state.error[f.id]&&this.state.error[f.id].length>0)||
             (!this.state.data[f.id]&&f.required))
         flag=f.placeholder
        });
        if(!flag)
        this.props.callback(this.state)
        else
         toast(`${flag} is a Mandatory Field`)
     }
    render(){
        let Modbody=null
        let {fields,action,formId}=this.props
        let result=fields.map((f,fi)=>{
        
        if(f.icon=='map')
            Modbody=<Modal style={{width:'50%',height:'50%',top:'10%',left:'30%'}}  onClose={()=>this.modalOpenClose('')} open={this.state.open=='map'}>
                <div className="F-map"style={{width:'100%'}} ><SeatMap admin select={(v)=>this.selectModalVal(v,f.name)}/></div></Modal>

        if(f.type=='text' || f.type=='checkbox' || f.type=='password' || f.type=='date'){
            return (<li key={'f'+fi}>
                    <label htmlFor={f.id}> {f.name}</label>
                    <input {...f} name={f.id} value={this.state.data[f.id]}  onChange={this.handleChange}/>
                    {f.icon?<AddBoxIcon style={{cursor:'pointer'}} onClick={()=>this.modalOpenClose(f.icon)}></AddBoxIcon>:null}
                </li>)
        }
        else if(f.type=='radio')
            {
            return (<li key={'f'+fi}><p>{f.name}</p><ul className="flex-inner">{
                f.radios.map((r,i)=>{
                    return ( <li key={'f'+fi+'-'+i}>
                        <input type="radio"  name={f.id} id={r.id} value={r.value}
                         checked={this.state.data[f.id]==r.value} onChange={this.handleChange} /> 
                         <label htmlFor={r.id} >{r.name}</label>
                         </li>)
                })
            }</ul></li>)
            }
        
    });

    return (
        <form id={formId} onSubmit={this.checkFields} autoComplete="off">
            <ul className="flex-outer">
            {result}
            {this.props.fields.map(x=>{
                if(this.state.error[x.id]){
                   return this.state.error[x.id].map((e,i)=>{
                        return <div className="error" key={i}>{e}</div>
                    })
                }
                return null;
            })}
            <li>  <button type="submit">{action}</button></li>
            </ul>
            {Modbody}
        </form>
        )
    }
    
}
ReactiveForm.propTypes={
    fields:PropTypes.array.isRequired,callback:PropTypes.func.isRequired,action:PropTypes.string.isRequired,
    formId:PropTypes.any
}

export default ReactiveForm