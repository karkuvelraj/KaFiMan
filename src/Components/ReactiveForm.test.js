import React from "react"
import {shallow} from "enzyme"
import ReactiveForm from './ReactiveForm'
import AddBoxIcon from '@material-ui/icons/AddBox';
let formFields=[
   {
        name:"dob" , type:"date",value:"",placeholder:"Date of birth" ,id:"dob"
    },{
        name:"passportNo" , type:"text",maxLength:10 ,value:"",placeholder:"Passport No" ,id:"passportNo", 
    },{
        name:"seatNo" , type:"text",maxLength:30 ,value:"",placeholder:"Seat No" ,id:"seatNo",readOnly:true,
        icon:'map'
    },{
        name:"flightNo" , type:"text",maxLength:30 ,value:"",placeholder:"Flight No" ,id:"flightNo", readOnly:true
    },{
        name:"Email Id" , type:"text",minLength:5,maxLength:40 ,value:"",placeholder:"Enter Email Id" ,id:"email",format:'email'
    }]

describe('check snapshot of form',()=>{
    it('four fields , one field with icon and one fields disabled',()=>{
        let callback=jest.fn();
        let label="save"
        let form=shallow(<ReactiveForm callback={callback} fields={formFields} action={label}/>)
        expect(form).toMatchSnapshot()
    })
})

describe('check form data changes',()=>{
  
    it('check button label from props',()=>{
        let callback=jest.fn();
        let label="save"
        let form=shallow(<ReactiveForm callback={callback} fields={formFields} action={label}/>)

        expect(form.find('button').text()).toBe(label)
    })
    it('should display icon end of input field ',()=>{
        let callback=jest.fn();
        let label="save"
        let form=shallow(<ReactiveForm callback={callback} fields={formFields} action={label}/>)

        expect(form.find(AddBoxIcon).exists()).toBeTruthy()
    })

    it('should display date field as first ',()=>{
        let callback=jest.fn();
        let label="save"
        let form=shallow(<ReactiveForm callback={callback} fields={formFields} action={label}/>)
        // console.log(form.html())
        expect(form.find('[type="date"]').length).toBe(1)
    })
        
    it('should call callback  on submit if no error',()=>{
        let callback=jest.fn();
        let label="save"
       
        let form=shallow(<ReactiveForm callback={callback} fields={formFields} action={label}/>)

        form.simulate('submit',{ target: {},preventDefault:jest.fn() })
        expect(callback).toBeCalled()
    })

    it('should display error on in valid email ',()=>{
        let callback=jest.fn();
        let label="save"

        let form=shallow(<ReactiveForm callback={callback} fields={formFields} action={label}/>)
 
        form.find('[id="email"]').simulate('change',{ target: {type:'text',name:'email',value:'fghfghj'} })
        
        expect(form.state().error['email'][0]).toBe('Email Id is not a valid email')
    })
})