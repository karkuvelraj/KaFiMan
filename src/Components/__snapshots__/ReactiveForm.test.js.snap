// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`check snapshot of form four fields , one field with icon and one fields disabled 1`] = `ShallowWrapper {}`;
